### DISCUSSION TITLE
Invariant Violation: useForm must be used within a form.

### DISCUSSION THREAD

**Discussion by danyhiol** [vazco/uniforms#903] [Q&A] [answered]
I tried to use the `useForm` hook in my component to replace the complicated access to error and others. But I keep getting the following error: `Invariant Violation: useForm must be used within a form`. What am I missing?
Thanks

**Comment by radekmie** [comment, marked as answer]
That'd mean your component using `useForm` is not within any form component, e.g., `AutoForm`. Could you check if that's the case?

**Reply by danyhiol** [reply]
Thank I could solve the issue by looking at the source code. I taught we could use `useForm` outside of the `form` itself to do things like validation easily without rendering a field inside the form itself. I finally went with `ref`.
Also, it would be nice if `uniforms` exports typing for the `formRef` the [doc](https://uniforms.tools/docs/api-forms/#methods) define some method already. But I could not find any place to import the typings from. Also, there is an `AutoFormProps`, but `AutoForm` itself (in my case antd) does not have typings. I may be missing something...

**Reply by radekmie** [reply]
I don't think there's a need to look into the source code - it clearly says, that it has to be used within the form. And yes, using ref to access it outside is correct.

When it comes to typing forms except the `BaseForm`, it's problematic due to the way how TypeScript treats generic mixins. We are aware of this issue, but thanks for this note.
